attestation-service/src/verifier/cca/mod.rs

```diff
@@ -0,0 +1,255 @@
+///     44235_1: h'00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
+///     44237_1: h'0476f988091be585ed41801aecfab858548c63057e16b0e676120bbd0d2f9c29e056c5d41a0130eb9c21517899dc23146b28e1b062bd3ea4b315fd219f1cbb528cb6e74ca49be16773734f61a1ca61031b2bbf3d918f2f94ffc4228e50919544ae',
+///     44236_1: "sha-256",
+///     44240_1: "sha-256",
+///     44238_1: h'75a1fbc79a7d20a5ff843b914dfd8093d40cd07dd633401c8c42d697be224801',
+///     44239_1: [
+///         h'0000000000000000000000000000000000000000000000000000000000000000',
+///         h'0000000000000000000000000000000000000000000000000000000000000000',
+///         h'0000000000000000000000000000000000000000000000000000000000000000',
+///         h'0000000000000000000000000000000000000000000000000000000000000000',
+///     ],
+/// }
+fn parse_cca_token(token: Vec<u8>) -> Result<Evidence> {
+    let evidence = Evidence {
+        cca_realm_delegated_token: RealmToken {
+            cca_realm_initial_measurement: "".to_string(),
+        },
+    };
+
+    // NOTE: For some corner case, the date cannot be parsed to token correctly but the token
+    // can be validated successfully by the CCA verifier e.g. `Err` value: Todo("Remaining bytes (00)")'
+    // Instead of throwing an error, just print it out in this case.
+    let mut di = match cbor_diag::parse_bytes(token) {
+        Ok(di) => di,
+        Err(err) => {
+            log::info!("Error: {:?}", err);
+            return Ok(evidence);
+        }
+    };
+
+    if let cbor_diag::DataItem::Tag {
+        tag: _,
+        bitwidth: _,
+        value,
+    } = di
+    {
+        di = *value;
+    }
+
+    if let cbor_diag::DataItem::Map { data, .. } = di {
+        for item in data {
+            if let cbor_diag::DataItem::ByteString(t) = item.1 {
+                let tok = cbor_diag::parse_bytes(t.data)?;
+                if let cbor_diag::DataItem::Tag { value, .. } = tok {
+                    if let cbor_diag::DataItem::Array { data, .. } = *value {
+                        if let cbor_diag::DataItem::ByteString(cose) = data
+                            .get(2)
+                            .ok_or_else(|| anyhow!("Cannot get raw bytes from token"))?
+                        {
+                            let v = &cose.data;
+                            match cbor_diag::parse_bytes(v) {
+                                Ok(claims) => {
+                                    info!("{}", claims.to_diag_pretty());
+                                }
+                                Err(e) => {
+                                    error!("Error parsing claims: {}", e);
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
```

Comment:
I'm not really a fan of nesting so many `if lef` statements. One way to adjust would be to make a new function that is called on each item. That function could return a result. Then you could just bail out of the function at each step if anything goes wrong. There might be other ways too.

It also seems like there is really only one error case here. What if there are no items in data, for instance (not sure if that is really possible). The function will return `Ok(evidence)` in that case. Is that correct?

Reply:
I will try to rewrite those code in my next day.

> The function will return Ok(evidence) in that case. Is that correct?

those piece of code is just a sample, if the plaintext of token could be returned after the verify by the veraison, then we can just remove those code.

> The function will return Ok(evidence) in that case. Is that correct?

yes, an empty evidence is return in this case.

Reply:
@fitzthum  updated to avoid too depth nesting here, not perfect but I hope it's better :)